core/src/main/java/com/linecorp/armeria/client/cookie/DefaultCookieJar.java

```diff
@@ -0,0 +1,167 @@
+import com.linecorp.armeria.common.CookieBuilder;
+import com.linecorp.armeria.common.Cookies;
+
+/**
+ * A default in-memory {@link CookieJar} implementation.
+ */
+final class DefaultCookieJar implements CookieJar {
+
+    private final Set<Cookie> store;
+    /**
+     * Used to find cookies for a host that matches a domain. For example, if there is a domain example.com,
+     * host example.com or foo.example.com will match all cookies in that entry.
+     */
+    private final Map<String, Set<Cookie>> filter;
+    private final CookiePolicy cookiePolicy;
+    private final ReentrantLock lock;
+
+    DefaultCookieJar() {
+        this(CookiePolicy.acceptOriginOnly());
+    }
+
+    DefaultCookieJar(CookiePolicy cookiePolicy) {
+        this.cookiePolicy = cookiePolicy;
+        store = new HashSet<>();
+        filter = new HashMap<>();
+        lock = new ReentrantLock();
+    }
+
+    @Override
+    public Cookies get(URI uri) {
+        requireNonNull(uri, "uri");
+        if (store.isEmpty()) {
+            return Cookies.of();
+        }
+        final String host = uri.getHost();
+        final String path = uri.getPath().isEmpty() ? "/" : uri.getPath();
+        final boolean secure = "https".equalsIgnoreCase(uri.getScheme());
+        final Set<Cookie> cookies = new HashSet<>();
+        lock.lock();
+        try {
+            filterGet(cookies, host, path, secure);
+        } finally {
+            lock.unlock();
+        }
+        return Cookies.of(cookies);
+    }
+
+    @Override
+    public void set(URI uri, Cookies cookies) {
+        requireNonNull(uri, "uri");
+        requireNonNull(cookies, "cookies");
+        lock.lock();
+        try {
+            for (Cookie cookie : cookies) {
+                cookie = ensureDomainAndPath(cookie, uri);
+                // remove similar cookie if present
+                store.remove(cookie);
+                if (!cookie.isExpired() && cookiePolicy.accept(uri, cookie)) {
+                    store.add(cookie);
+                    final Set<Cookie> cookieSet = filter.computeIfAbsent(cookie.domain(), s -> new HashSet<>());
+                    // remove similar cookie if present
+                    cookieSet.remove(cookie);
+                    cookieSet.add(cookie);
+                }
+            }
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Ensures this cookie has domain and path attributes, otherwise sets them to default values. If domain
+     * is absent, the default is the request host, with {@code host-only} flag set to {@code true}. If path is
+     * absent, the default is computed from the request path. See RFC 6265
+     * <a href="https://tools.ietf.org/html/rfc6265#section-5.3">5.3</a> and
+     * <a href="https://tools.ietf.org/html/rfc6265#section-5.1.4">5.1.4</a>
+     */
+    @VisibleForTesting
+    Cookie ensureDomainAndPath(Cookie cookie, URI uri) {
+        final CookieBuilder cb = cookie.toBuilder();
```

Comment:
If a `cookie` has a valid `domain` and `path`, how about returning `cookie` itself without an additional object creation.